sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala

```diff
@@ -1213,6 +1213,71 @@ case class ToUTCTimestamp(left: Expression, right: Expression)
 }
 
 /**
+ * This modifies a timestamp to show how the display time changes going from one timezone to
+ * another, for the same instant in time.
+ *
+ * We intentionally do not provide an ExpressionDescription as this is not meant to be exposed to
+ * users, its only used for internal conversions.
+ */
+private[spark] case class TimestampTimezoneCorrection(
```

Comment:
do we need a whole expression for this? can't we just reuse existing expressions? It's just simple arithmetics isn't it?

Reply:
I guess you could use ToUTCTimestamp / FromUTCTimestamp for this, but that would be more expensive since you'd be doing the conversion twice for each value.

Reply:
What I'm saying is the analysis rule can just determine the delta, and then just do a simple add/delete.

Reply:
Hmm, let me see if I can figure that out.

Reply:
Unfortunately the offset depends on the actual date, so a timezone conversion can not be simplified to a simple delta.

Daylight saving time starts and ends on different days in different timezones, while some timezones don't have DST changes at all. Additionally, timezone rules have changed over time and keep changing. Both the basic timezone offset and the DST rules themselves could change (and have changed) over time.

Reply:
> Additionally, timezone rules have changed over time and keep changing. 

Ah, yes, that makes sense... it's also why my initial tests were failing at the DST boundaries. :-/